### PR TITLE
feat: adds QueryOptions and Custom headers

### PR DESCRIPTION
**Pull request by karel-rehor**
## Proposed Changes
   * Wrap extra query arguments into a QueryOptions structure
   * Add custom headers to queries
   * On merge of option structures before using transport calls - instead of overwriting values, merge them from...
      * `client_options.headers`
      * `client_options.writeOptions` or `client_options.queryOptions`
      * options arguments passed in `query` or `write` method calls 
   * when `client_options.queryOptions` are defined in the client constructor, these become default when making a query without adding `QueryOptions` to the query call.  This makes it possible for example to define `influxql` as a default queryType for all calls (e,g `ClientOptions{ ... queryOptions:{ type: 'influxql' } }`), and then to not have to add `QueryOptions { type: 'influxql'}` to each call.
   * Update current examples to use `QueryOptions`
   * For tests adds a `MockService` and `TestServer` based on `Flight.gRPC.server` which can be generated in scripts.
   * Update documentation

## Checklist

- [x] CHANGELOG.md updated
- [x] Rebased/mergeable
- [x] A test has been added if appropriate
- [x]  Tests pass
- [x] Commit messages are [conventional](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Sign [CLA](https://www.influxdata.com/legal/cla/) (if not already signed)

### CODE REVIEW

File: packages/client/src/InfluxDBClient.ts [hunk 1]
```diff
@@ -135,23 +161,32 @@ export default class InfluxDBClient {
    *
    * @param query - The query string.
    * @param database - The name of the database to query.
-   * @param queryType - The type of query (default: 'sql').
-   * @param namedParams - for sql queries parameters to be used
+   * @param queryOptions - The options for the query (default: \{ type: 'sql' \}).
+   * @example
+   * ```typescript
+   *    client.query('SELECT * from testData', 'CI_TEST', {
```
Comment (bednar) [inline packages/client/src/InfluxDBClient.ts:167]: Maybe we can use something more "production" in examples like:

```suggestion
   *    client.query('SELECT * from cpu', 'performance_db', {
```

File: packages/client/test/TestServer.ts [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+  error(message: string, error) {
+    consoleLogger.error(message, error)
+  },
+  warn(message: string, error) {
+    consoleLogger.warn(message, error)
+  },
+})
+
+// todo - when run in parallel can get conflicts?
+// so may need to increment port number or isolate runs
```
Comment (bednar) [inline packages/client/test/TestServer.ts:20]: Fix this TODO or remove this comment from sources.

File: packages/client/test/unit/Influxdb.test.ts [hunk 3]
```diff
@@ -43,20 +44,24 @@ describe('InfluxDB', () => {
...
     client.query(query)
 
-    expect(queryStub.calledOnceWith(query, database, 'sql')).to.be.true
+    // expect(queryStub.calledOnceWith(query, database, 'sql')).to.be.true
+    expect(queryStub.calledOnceWith(query, database)).to.be.true
     queryStub.resetHistory()
 
     client.query(query, 'another')
-    expect(queryStub.calledOnceWith(query, 'another', 'sql')).to.be.true
+    // expect(queryStub.calledOnceWith(query, 'another', 'sql')).to.be.true
```
Comment (bednar) [inline packages/client/test/unit/Influxdb.test.ts:52]: Please remove this commented line.
```suggestion
```

File: packages/client/test/unit/Influxdb.test.ts [hunk 4]
```diff
@@ -43,20 +44,24 @@ describe('InfluxDB', () => {
...
+    expect(queryStub.calledOnceWith(query, 'another')).to.be.true
 
     // queryPoints
     client.queryPoints(query)
 
-    expect(queryPointsStub.calledOnceWith(query, database, 'sql')).to.be.true
+    // expect(queryPointsStub.calledOnceWith(query, database, 'sql')).to.be.true
```
Comment (bednar) [inline packages/client/test/unit/Influxdb.test.ts:58]: Please remove this commented line.
```suggestion
```

File: packages/client/test/unit/Influxdb.test.ts [hunk 5]
```diff
@@ -43,20 +44,24 @@ describe('InfluxDB', () => {
...
+    expect(queryPointsStub.calledOnceWith(query, database)).to.be.true
     queryPointsStub.resetHistory()
 
     client.queryPoints(query, 'another')
-    expect(queryPointsStub.calledOnceWith(query, 'another', 'sql')).to.be.true
+    // expect(queryPointsStub.calledOnceWith(query, 'another', 'sql')).to.be.true
```
Comment (bednar) [inline packages/client/test/unit/Influxdb.test.ts:63]: Please remove this commented line.
```suggestion
```

**Review by bednar — CHANGES_REQUESTED** [review]
Thanks for your awesome PR! 👍 

There are only a few requirements:

File: packages/client/src/InfluxDBClient.ts (hunk 1 above)
Comment (karel-rehor) [reply packages/client/src/InfluxDBClient.ts]: I've updated examples in TS doc to reflect production like values

File: packages/client/test/unit/Influxdb.test.ts (hunk 3 above)
Comment (karel-rehor) [reply packages/client/test/unit/Influxdb.test.ts]: removed

File: packages/client/test/unit/Influxdb.test.ts (hunk 4 above)
Comment (karel-rehor) [reply packages/client/test/unit/Influxdb.test.ts]: removed

File: packages/client/test/unit/Influxdb.test.ts (hunk 5 above)
Comment (karel-rehor) [reply packages/client/test/unit/Influxdb.test.ts]: removed

File: packages/client/test/TestServer.ts (hunk 2 above)
Comment (karel-rehor) [reply packages/client/test/TestServer.ts:20]: I created a temporary script to check behavior in a parallel run and then ran it.  I also tried the same for  `test:unit`
```
"test:integration:parallel": "mocha 'test/integration/**/*.test.ts' --parallel --exit",
```
I did not see any negative side effects - for now.  In fact parallel runs were a second or two slower.  The project is small enough that test parallelization brings little added benefit.  

I've removed this `todo` comment.  

Still, I think there are some ways in which this TestServer can be improved.  However,  at the moment they are of a lower priority.

**Review by bednar — COMMENTED** [review]
LGTM 🚀

**Review by bednar — APPROVED** [review]

**Pull request merged**